presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -2418,6 +2425,39 @@ private static Domain buildColumnDomain(ColumnHandle column, List<HivePartition>
         return Optional.of(new ConnectorNewTableLayout(partitioningHandle, partitionColumns));
     }
 
+    @Override
+    public Optional<ConnectorNewTableLayout> getPreferredShuffleLayoutForInsert(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        HiveTableHandle hiveTableHandle = (HiveTableHandle) tableHandle;
+        SchemaTableName tableName = hiveTableHandle.getSchemaTableName();
+        Table table = metastore.getTable(tableName.getSchemaName(), tableName.getTableName())
+                .orElseThrow(() -> new TableNotFoundException(tableName));
+
+        Optional<HiveBucketHandle> hiveBucketHandle = getHiveBucketHandle(table);
+        if (hiveBucketHandle.isPresent()) {
+            // For bucketed table, table partitioning (i.e. the bucketing scheme) should be respected,
+            // and there is no additional preferred shuffle partitioning
+            return Optional.empty();
+        }
+
+        if (!isShufflePartitionedColumnsForTableWriteEnabled(session) || table.getPartitionColumns().isEmpty()) {
+            return Optional.empty();
+        }
+
+        // TODO: the shuffle partitioning doesn't have to be the same as Hive bucket partitioning
```

Comment:
@wenleix I don't understand this TODO. It seems to me that bucketed tables are handled above and would never reach that code. Would you clarify?

Reply:
@mbasmanova : Sorry for the confusion, I mean we don't have to use `HivePartitionHandle` for the shuffle partitioning. Thinking about implementing a different `HiveShufflePartitioningHandle` that distribute the keys more uniformly (Hive bucket function is not that great :) )